server/app/views/admin/programs/ProgramBlockEditView.java

```diff
@@ -14,140 +13,244 @@
 import controllers.admin.routes;
 import forms.BlockForm;
 import j2html.TagCreator;
+import j2html.tags.DomContent;
 import j2html.tags.specialized.ButtonTag;
 import j2html.tags.specialized.DivTag;
 import j2html.tags.specialized.FormTag;
 import j2html.tags.specialized.InputTag;
+import java.util.ArrayList;
 import java.util.Optional;
 import java.util.OptionalLong;
-import java.util.stream.IntStream;
 import play.mvc.Http.HttpVerbs;
 import play.mvc.Http.Request;
 import play.twirl.api.Content;
 import services.program.BlockDefinition;
 import services.program.ProgramDefinition;
 import services.program.ProgramDefinition.Direction;
 import services.program.ProgramQuestionDefinition;
-import services.program.predicate.PredicateDefinition;
 import services.question.types.QuestionDefinition;
 import services.question.types.StaticContentQuestionDefinition;
 import views.HtmlBundle;
 import views.ViewUtils;
-import views.admin.AdminLayout;
-import views.admin.AdminLayout.NavPage;
+import views.ViewUtils.BadgeStatus;
 import views.admin.AdminLayoutFactory;
 import views.components.FieldWithLabel;
 import views.components.Icons;
 import views.components.Modal;
 import views.components.QuestionBank;
-import views.components.SvgTag;
 import views.components.ToastMessage;
 import views.style.AdminStyles;
 import views.style.ReferenceClasses;
 import views.style.StyleUtils;
 
-/** Renders a page for an admin to edit the configuration for a single block of a program. */
-public final class ProgramBlockEditView extends ProgramBlockView {
+/**
+ * Renders a page for an admin to edit the configuration for program, including a single block of a
+ * program. A block is a synonym for a Screen. The ProgramBlockEditView is very similar to the
+ * ProgramBlockReadOnlyView, but specifically adds all UI functionality that is needed for editing.
+ */
+public final class ProgramBlockEditView extends ProgramBlockReadOnlyView {
 
-  private final AdminLayout layout;
-  private final boolean featureFlagOptionalQuestions;
-
-  public static final String ENUMERATOR_ID_FORM_FIELD = "enumeratorId";
-  public static final String MOVE_QUESTION_POSITION_FIELD = "position";
   private static final String CREATE_BLOCK_FORM_ID = "block-create-form";
   private static final String CREATE_REPEATED_BLOCK_FORM_ID = "repeated-block-create-form";
   private static final String DELETE_BLOCK_FORM_ID = "block-delete-form";
 
+  private final boolean featureFlagOptionalQuestions;
+  private InputTag csrfTag;
+  private ArrayList<Modal> modals;
```

Comment:
After our conversation yesterday I had a look to see if the solution you suggested could work. To summarize, the suggestion was to have the modals be a local variable in one of the methods in the parent class inthe interest of it getting destroyed each time. The partent class would create it and document its purpose and the child class could add to it so that modals can be added to the html bundle by the parent.

However, this got kind of messy and reminded me of the initial reason why I introduced the state in the child at all. The challenge is (as you point out in your comment) that the modal is needed in two places:
- when the bundle is created (to add it) in createHtmlBundle()
- to get the button and allow adding it to the UI in prepareContentForBlockPanel

If I create the modals list in the parent locally, it has to be passed through various methods including the methods that are specific to where exactly the child wants to use the modal. In other words, if another child wanted to add a modal in a different place, the code would not be easily reusable. This version therefore ads a lot of logic specific to one child implementation in the parent. it also gets less readable overall. Overall it didn't look right.

Here is what I did instead: I documented the methods that are overriden by the child better to provide a contract that they are called exactly once per render so that the child can do its implementation based on that contract.
I use a synchronized list as discussed and recreate the list for each new render to avoid concurrency issues and add an extra safety measure that helps avoid modal duplication.
There is the additional option of ensuring in the child that no duplication exists in the list in case the method creating the modal does get called twice. As that would be breaking the documented contract, I didn't implement it, but could add it.

What do you think?